import.go

```diff
@@ -109,54 +101,131 @@ func importYamlWithTemplateFunc(r io.Reader, tmplFunc TemplateFunc) ([]*YamlSect
 		}}
 	}
 	for i := range chunks {
-		vp := viper.New()
-		vp.SetConfigType("yaml")
+		//Each chunk is a document, needs to be filtered so the consumer gets the sections they want
+		// then run the supplied template, then Vipered
+		// this avoids running templates on unrelated sections that we may not have the data for
+
+		//extract section
 		var b []byte
 		if i < len(chunks)-1 {
 			b = data[chunks[i][1]:chunks[i+1][0]]
 		} else {
 			b = data[chunks[i][1]:]
 		}
-		if err := vp.ReadConfig(bytes.NewBuffer(b)); err != nil {
-			fmt.Printf("---\n%v\n", string(b))
-			return nil, errors.Wrap(err, "failed to import yaml section")
-		}
-		sections = append(sections, &YamlSection{Bytes: b, Viper: vp})
+
+		//save completed section
+		sections = append(sections, &YamlSection{
+			Bytes:         b,
+			OriginalBytes: b,
+		})
 	}
 	return sections, nil
 }
 
-func importYaml(r io.Reader) ([]*YamlSection, error) {
-	sections := []*YamlSection{}
-	buf := new(bytes.Buffer)
-	_, err := buf.ReadFrom(r)
+func Filter(in []*YamlSection, filters []string) ([]*YamlSection, error) {
+	ret := []*YamlSection{}
+	for _, section := range in {
+		//Each section is a document, needs to be filtered so the consumer gets the sections they want
+		// then run the supplied template, then Vipered
+		// this avoids running templates on unrelated sections that we may not have the data for
+
+		//run filters
+		if !func() bool {
+			for _, v := range filters {
+				rx := regexp.MustCompile(v)
+				scanner := bufio.NewScanner(bytes.NewBuffer(section.OriginalBytes))
+				for scanner.Scan() {
+					if rx.MatchString(scanner.Text()) {
+						return true
+					}
+				}
+			}
+			return false
+		}() {
+			continue
+		}
+		//save completed section
+		ret = append(ret, section)
+	}
+	return ret, nil
+}
+
+func (yp *Yp) ApplyFilters(s string, filters []string) error {
+	if _, ok := yp.Files[s]; !ok {
+		return errors.WithFields(errors.Fields{
+			"File": s,
+		}).New("Apply filters failed, no such file loaded")
+	}
+	out, err := Filter(yp.Files[s], filters)
 	if err != nil {
-		return nil, fmt.Errorf(fmt.Sprintf("could not read data %v", err))
+		return err
 	}
-	data := buf.Bytes()
-	rxChunks := regexp.MustCompile(`---`)
-	chunks := rxChunks.FindAllIndex(data, -1)
-	if len(chunks) == 0 {
-		//This is missing the end delimiter ('---') or both,
-		//either way we are treating the whole file as 1 section
-		chunks = [][]int{[]int{
-			int(0), int(len(data)),
-		}}
+	yp.Files[s] = out
+	return nil
+}
+
+func (yp *Yp) applyNullTemplate(name string) error {
+	if _, ok := yp.Files[name]; !ok {
+		return errors.WithFields(errors.Fields{"Name": name}).New("File has not been imported")
 	}
-	for i := range chunks {
-		vp := viper.New()
-		vp.SetConfigType("yaml")
-		var b []byte
-		if i < len(chunks)-1 {
-			b = data[chunks[i][1]:chunks[i+1][0]]
-		} else {
-			b = data[chunks[i][1]:]
+
+	tf := func(in []byte) ([]byte, error) {
```

Comment:
A lot of use of anonymous functions. This will make the unit tests less granular and a little more messing.